plugin/core/protocol.py

```diff
@@ -6335,8 +6280,9 @@ def to_lsp(self) -> Position:
     'session_name': str
 })
 
-CompletionItemDefaults = __CompletionList_itemDefaults_Type_1
-CompletionEditRange = __CompletionList_itemDefaults_editRange_Type_1
-
 # Temporary for backward compatibility with LSP packages.
 RangeLsp = Range
+
+
+def is_text_edit(value: Any) -> TypeGuard[TextEdit]:
+    return isinstance(value, dict) and bool(value.get('range')) and bool(value.get('newText'))
```

Comment:
yeah, that's not correct. Use `in` to check for existance of property and not `bool`

Reply:
But also this does not exclude `AnnotatedTextEdit` if that was your goal. You'd need to explicitly check for lack of `annotationId`

Reply:
Also true, I did the filtering to make pyright happy. There is probably a better way to handle this, than the way I did.

Reply:
Then we can probably just extend the type of `changes` so that it allows `AnnotatedTextEdit`. We are not supposed to get them anyway unless we say so.

Reply:
> yeah, that's not correct. Use in to check for existance of property and not bool

Thanks, this is the reason for this bug reported here https://github.com/sublimelsp/LSP/pull/2500#discussion_r1666795122

Reply:
I tried to extend the type:
```diff
- WorkspaceChanges = dict[str, tuple[list[TextEdit], Optional[int]]]
+ WorkspaceChanges = dict[str, tuple[list[TextEdit | AnnotatedTextEdit | SnippetTextEdit], Optional[int]]]
```

I quit from changing the code because:
- there is this error which I cant seem to figure out how to solve, and not cause type issues on other places  -> 
```
// Packages/LSP/plugin/core/edit.py:
32:17  error   Argument of type "tuple[List[TextEdit], None]" cannot be assigned to parameter "value" of type "tuple[list[TextEdit | AnnotatedTextEdit | SnippetTextEdit], int | None]" in function "__setitem__" ​Pyright
                    "tuple[List[TextEdit], None]" is incompatible with "tuple[list[TextEdit | AnnotatedTextEdit | SnippetTextEdit], int | None]"
                      Tuple entry 1 is incorrect type
                        "List[TextEdit]" is incompatible with "list[TextEdit | AnnotatedTextEdit | SnippetTextEdit]"
                          Type parameter "_T@list" is invariant, but "TextEdit" is not the same as "TextEdit | AnnotatedTextEdit | SnippetTextEdit"
                          Consider switching from "list" to "Sequence" which is covariant
```

AnnotatedTextEdit don't require code changes, however SnippetTextEdit do require code changes because they doesn't have `newText`, instead they have a `snippet` property. I rather not extend the PR to support SnippetTextEdit, at least not in this PR. I would leave the filtering of Text edits `[edit for edit in edits if is_text_edit(edit)]`. I just wanted to update the types with this PR.

Reply:
`AnnotatedTextEdit` is defined as `interface AnnotatedTextEdit extends TextEdit` so it does include `newText`, same as `TextEdit`.

I can look into the error later.

Reply:
I've pushed a change that should address it in a different way

Reply:
Thanks, I like the way it was addressed